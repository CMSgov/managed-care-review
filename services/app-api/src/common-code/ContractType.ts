import type {
    Contract,
    ContractPackageSubmission,
    ContractRevision,
} from '../gen/gqlServer'

const getLastContractSubmission = (
    contract: Contract
): ContractPackageSubmission | undefined => {
    return (
        (contract.packageSubmissions && contract.packageSubmissions[0]) ??
        undefined
    )
}

const getContractRev = (contract: Contract): ContractRevision | undefined => {
    if (contract.draftRevision) {
        return contract.draftRevision
    } else {
        return getLastContractSubmission(contract)?.contractRevision
    }
}
const isContractOnly = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_ONLY'
}

const isBaseContract = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'BASE'
}

const isContractAmendment = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'AMENDMENT'
}

const isCHIPOnly = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.populationCovered === 'CHIP'
}

const isContractAndRates = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_AND_RATES'
}

const isContractWithProvisions = (contract: Contract): boolean =>
    isContractAmendment(contract) ||
    (isBaseContract(contract) && !isCHIPOnly(contract))

const isSubmitted = (contract: Contract): boolean =>
    contract.status === 'SUBMITTED'

export {
    isContractWithProvisions,
    isBaseContract,
    isContractAmendment,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isSubmitted,
}
