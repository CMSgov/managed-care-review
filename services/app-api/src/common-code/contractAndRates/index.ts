export {
    isBaseContract,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isContractAmendment,
    isSubmitted,
    isContractWithProvisions,
    getLastContractSubmission,
} from './contractData'

export {
    getProvisionDictionary,
    sortModifiedProvisions,
    generateApplicableProvisionsList,
    generateProvisionLabel,
    isMissingProvisions,
    hasValidModifiedProvisions,
} from './ContractProvisions'

export {
    modifiedProvisionMedicaidBaseKeys,
    modifiedProvisionMedicaidAmendmentKeys,
    provisionCHIPKeys,
    isCHIPProvision,
    isMedicaidAmendmentProvision,
    isMedicaidBaseProvision,
} from './ModifiedProvisions'
