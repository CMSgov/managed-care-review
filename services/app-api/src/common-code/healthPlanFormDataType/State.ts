type StateType = {
    code: string
    name: string
    programs: ProgramArgType[]
}

interface ProgramArgType {
    id: string
    name: string // short name. This is used most often in the application including in submission name
    fullName: string // full name is used in submission summary page
    isRateProgram: boolean // specifies if program relates to rates rather than contract
}

export type { StateType, ProgramArgType }
