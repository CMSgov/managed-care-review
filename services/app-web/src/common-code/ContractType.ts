import { Contract, ContractRevision, UnlockedContract } from '../gen/gqlClient'
import { getLastContractSubmission } from '../gqlHelpers/contractsAndRates'

const getContractRev = (contract: Contract | UnlockedContract): ContractRevision | undefined => {
    if (contract.draftRevision) {
        return contract.draftRevision
    } else {
        return getLastContractSubmission(contract)?.contractRevision
    }
}
const isContractOnly = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_ONLY' 
}


const isBaseContract = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'BASE'
}

const isContractAmendment = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'AMENDMENT'
}

const isCHIPOnly = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.populationCovered === 'CHIP'
}

const isContractAndRates = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_AND_RATES'
}

const isContractWithProvisions = (contract: Contract | UnlockedContract): boolean =>
    isContractAmendment(contract) || (isBaseContract(contract) && !isCHIPOnly(contract))

const isSubmitted = (contract: Contract | UnlockedContract): boolean =>
    contract.status === 'SUBMITTED'

export {
    isContractWithProvisions,
    isBaseContract,
    isContractAmendment,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isSubmitted,
}
