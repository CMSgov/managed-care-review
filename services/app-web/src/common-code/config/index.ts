export { assertIsAuthMode, assertNever } from './config'
export type { AuthModeType } from './config'
