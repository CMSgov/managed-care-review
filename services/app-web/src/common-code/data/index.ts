export {removeNullAndUndefined, replaceNullsWithUndefineds } from './dataUtilities'