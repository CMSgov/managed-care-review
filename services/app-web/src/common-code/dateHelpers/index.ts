export { formatCalendarDate, formatRateNameDate } from './calendarDate'
export { formatGQLDate } from './gqlDate'
export { dayjs } from './dayjs'
