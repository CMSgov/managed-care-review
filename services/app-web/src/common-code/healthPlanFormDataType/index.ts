// TEMPORARY: these files are embedded in app-web for now b/c
// CRA prevents you from importing code outside of /src
// The fix is to use yarn workspaces to allow us to import shared packages
// Domain Types
export type {
    ActuarialFirmType,
    ActuaryCommunicationType,
    ActuaryContact,
    ContractAmendmentInfo,
    ContractExecutionStatus,
    ContractType,
    DocumentCategoryType,
    UnlockedHealthPlanFormDataType,
    ManagedCareEntity,
    RateType,
    RateCapitationType,
    StateContact,
    SubmissionDocument,
    SubmissionType,
    RateInfoType,
    PopulationCoveredType,
    SharedRateCertDisplay,
} from './UnlockedHealthPlanFormDataType'

export type {
    FederalAuthority,
    CHIPFederalAuthority,
} from './FederalAuthorities'

export type {
    ModifiedProvisionsMedicaidAmendment,
    ModifiedProvisionsMedicaidBase,
    CHIPModifiedProvisions,
    GeneralizedProvisionType,
    MedicaidAmendmentProvisionType,
    GeneralizedModifiedProvisions,
} from './ModifiedProvisions'

export {
    federalAuthorityKeys,
    federalAuthorityKeysForCHIP,
} from './FederalAuthorities'

export {
    modifiedProvisionMedicaidBaseKeys,
    modifiedProvisionMedicaidAmendmentKeys,
    provisionCHIPKeys,
    isCHIPProvision,
    isMedicaidAmendmentProvision,
} from './ModifiedProvisions'

export type { LockedHealthPlanFormDataType } from './LockedHealthPlanFormDataType'
export {
    hasValidContract,
    hasValidDocuments,
    hasValidRates,
    hasAnyValidRateData,
    hasValidSupportingDocumentCategories,
    isContractAndRates,
    isContractOnly,
    isUnlockedHealthPlanFormData,
    isLockedHealthPlanFormData,
    isSubmitted,
    programNames,
    packageName,
    generateRateName,
    convertRateSupportingDocs,
    removeInvalidProvisionsAndAuthorities,
    removeRatesData,
    hasValidPopulationCoverage,
    isValidAndCurrentLockedHealthPlanFormData,
} from './healthPlanFormData'

export type { HealthPlanFormDataType } from './HealthPlanFormDataType'
export type { ProgramArgType, StateType } from './State'

export { StateCodes, isValidStateCode } from './StateCodeType'
export type { StateCodeType } from './StateCodeType'
