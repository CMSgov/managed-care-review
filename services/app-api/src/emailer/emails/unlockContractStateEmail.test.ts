import {
    testEmailConfig,
    mockMNState,
    mockContractRev,
} from '../../testHelpers/emailerHelpers'
import { unlockContractStateEmail } from './index'
import { packageName } from '../../common-code/healthPlanFormDataType'

const unlockData = {
    updatedBy: {
        email: 'josh@example.com',
        role: 'CMS_USER' as const,
        givenName: 'Josh',
        familyName: 'Knope',
    },
    updatedAt: new Date('02/01/2022'),
    updatedReason: 'Adding rate certification.',
}

const defaultStatePrograms = mockMNState().programs
const defaultSubmitters = ['test1@example.com', 'test2@example.com']

test('subject line is correct and clearly states submission is unlocked', async () => {
    const sub = mockContractRev()

    const name = packageName(
        sub.contract.stateCode,
        sub.contract.stateNumber,
        sub.formData.programIDs,
        defaultStatePrograms
    )
    const template = await unlockContractStateEmail(
        sub,
        unlockData,
        testEmailConfig(),
        defaultStatePrograms,
        defaultSubmitters
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was unlocked by CMS`),
        })
    )
})

test('includes expected data summary for a contract and rates submission unlock State email', async () => {
    const sub = mockContractRev()

    const template = await unlockContractStateEmail(
        sub,
        unlockData,
        testEmailConfig(),
        defaultStatePrograms,
        defaultSubmitters
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(/Unlocked by: josh/),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(/Unlocked on: 02/),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(
                /Reason for unlock: Adding rate certification./
            ),
        })
    )

    expect(template.bodyText?.match(/Rate Cert Name/)?.length).toBe(1)
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.rateRevisions[0].formData.rateCertificationName ?? ''
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(
                /You must revise the submission before CMS can continue reviewing it/
            ),
        })
    )
})

test('includes expected data summary for a multi-rate contract and rates submission unlock State email', async () => {
    const sub = mockContractRev({
        rateRevisions: [
            {
                id: '12345',
                rateID: '6789',
                submitInfo: undefined,
                unlockInfo: undefined,
                createdAt: new Date(11 / 27 / 2023),
                updatedAt: new Date(11 / 27 / 2023),
                formData: {
                    id: 'test-id-1234',
                    rateID: 'test-id-1234',
                    rateType: 'NEW',
                    rateCapitationType: 'RATE_CELL',
                    rateDocuments: [
                        {
                            s3URL: 's3://bucketname/key/test1',
                            name: 'foo',
                            sha256: 'fakesha',
                            dateAdded: new Date(11 / 27 / 2023),
                        },
                    ],
                    supportingDocuments: [],
                    rateDateStart: new Date('01/01/2024'),
                    rateDateEnd: new Date('01/01/2025'),
                    rateDateCertified: new Date('01/01/2024'),
                    amendmentEffectiveDateStart: new Date('01/01/2024'),
                    amendmentEffectiveDateEnd: new Date('01/01/2025'),
                    rateProgramIDs: ['3fd36500-bf2c-47bc-80e8-e7aa417184c5'],
                    rateCertificationName: 'Rate Cert Name 1',
                    certifyingActuaryContacts: [
                        {
                            actuarialFirm: 'DELOITTE',
                            name: 'Actuary Contact 1',
                            titleRole: 'Test Actuary Contact 1',
                            email: 'actuarycontact1@example.com',
                        },
                    ],
                    addtlActuaryContacts: [],
                    actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                    packagesWithSharedRateCerts: [
                        {
                            packageName: 'pkgName',
                            packageId: '12345',
                            packageStatus: 'SUBMITTED',
                        },
                    ],
                },
            },
            {
                id: '12345',
                rateID: '6789',
                submitInfo: undefined,
                unlockInfo: undefined,
                createdAt: new Date(11 / 27 / 2023),
                updatedAt: new Date(11 / 27 / 2023),
                formData: {
                    id: 'test-id-1234',
                    rateID: 'test-id-1234',
                    rateType: 'NEW',
                    rateCapitationType: 'RATE_CELL',
                    rateDocuments: [
                        {
                            s3URL: 's3://bucketname/key/test1',
                            name: 'foo',
                            sha256: 'fakesha',
                            dateAdded: new Date(11 / 27 / 2023),
                        },
                    ],
                    supportingDocuments: [],
                    rateDateStart: new Date('01/01/2024'),
                    rateDateEnd: new Date('01/01/2025'),
                    rateDateCertified: new Date('01/01/2024'),
                    amendmentEffectiveDateStart: new Date('01/01/2024'),
                    amendmentEffectiveDateEnd: new Date('01/01/2025'),
                    rateProgramIDs: ['3fd36500-bf2c-47bc-80e8-e7aa417184c5'],
                    rateCertificationName: 'Rate Cert Name 2',
                    certifyingActuaryContacts: [
                        {
                            actuarialFirm: 'DELOITTE',
                            name: 'Actuary Contact 1',
                            titleRole: 'Test Actuary Contact 1',
                            email: 'actuarycontact1@example.com',
                        },
                    ],
                    addtlActuaryContacts: [],
                    actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                    packagesWithSharedRateCerts: [
                        {
                            packageName: 'pkgName',
                            packageId: '12345',
                            packageStatus: 'SUBMITTED',
                        },
                    ],
                },
            },
            {
                id: '12345',
                rateID: '6789',
                submitInfo: undefined,
                unlockInfo: undefined,
                createdAt: new Date(11 / 27 / 2023),
                updatedAt: new Date(11 / 27 / 2023),
                formData: {
                    id: 'test-id-1234',
                    rateID: 'test-id-1234',
                    rateType: 'NEW',
                    rateCapitationType: 'RATE_CELL',
                    rateDocuments: [
                        {
                            s3URL: 's3://bucketname/key/test1',
                            name: 'foo',
                            sha256: 'fakesha',
                            dateAdded: new Date(11 / 27 / 2023),
                        },
                    ],
                    supportingDocuments: [],
                    rateDateStart: new Date('01/01/2024'),
                    rateDateEnd: new Date('01/01/2025'),
                    rateDateCertified: new Date('01/01/2024'),
                    amendmentEffectiveDateStart: new Date('01/01/2024'),
                    amendmentEffectiveDateEnd: new Date('01/01/2025'),
                    rateProgramIDs: ['3fd36500-bf2c-47bc-80e8-e7aa417184c5'],
                    rateCertificationName: 'Rate Cert Name 3',
                    certifyingActuaryContacts: [
                        {
                            actuarialFirm: 'DELOITTE',
                            name: 'Actuary Contact 1',
                            titleRole: 'Test Actuary Contact 1',
                            email: 'actuarycontact1@example.com',
                        },
                    ],
                    addtlActuaryContacts: [],
                    actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                    packagesWithSharedRateCerts: [
                        {
                            packageName: 'pkgName',
                            packageId: '12345',
                            packageStatus: 'SUBMITTED',
                        },
                    ],
                },
            },
        ],
    })
    const template = await unlockContractStateEmail(
        sub,
        unlockData,
        testEmailConfig(),
        defaultStatePrograms,
        defaultSubmitters
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(/Unlocked by: josh/),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(/Unlocked on: 02/),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(
                /Reason for unlock: Adding rate certification./
            ),
        })
    )
    //Expect only have 3 rate names using regex to match name pattern specific to rate names.
    expect(template.bodyText?.match(/Rate Cert Name/g)?.length).toBe(3)
    //First Rate certification
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.rateRevisions[0].formData.rateCertificationName ?? ''
            ),
        })
    )
    //Second Rate certification
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.rateRevisions[1].formData.rateCertificationName ?? ''
            ),
        })
    )
    //Third Rate certification
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                sub.rateRevisions[2].formData.rateCertificationName ?? ''
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(
                /You must revise the submission before CMS can continue reviewing it/
            ),
        })
    )
})

test('does includes the correct submission URL', async () => {
    const sub = mockContractRev()
    const template = await unlockContractStateEmail(
        sub,
        unlockData,
        testEmailConfig(),
        defaultStatePrograms,
        defaultSubmitters
    )

    if (template instanceof Error) {
        console.error(template)
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                `http://localhost/submissions/12345/edit/review-and-submit`
            ),
        })
    )
})

test('renders overall email as expected', async () => {
    const sub = mockContractRev()
    const template = await unlockContractStateEmail(
        sub,
        unlockData,
        testEmailConfig(),
        defaultStatePrograms,
        defaultSubmitters
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template.bodyHTML).toMatchSnapshot()
})
