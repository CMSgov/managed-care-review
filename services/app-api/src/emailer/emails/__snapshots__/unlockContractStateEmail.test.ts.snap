// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders overall email as expected 1`] = `
"Submission MCR-MN-0004-SNBC was unlocked by CMS<br />
<br />
<b>Unlocked by:</b> josh@example.com<br />
<b>Unlocked on:</b> 01/31/2022<br />
<b>Reason for unlock:</b> Adding rate certification.<br />
<b>Rate name</b>: Rate Cert Name<br />
<br />
<b>You must revise the submission before CMS can continue reviewing it.<br />
<br />
<a href="http://localhost/submissions/12345/edit/review-and-submit">Open the submission in MC-Review to make edits.</a>
"
`;
